### PR TITLE
Add EvictOldest to public API, add OnEviction and OnExpiration for setting callbacks outside of the Config

### PR DESCRIPTION
**Pull request by danielstjules**
Minimal changes, not breaking the api this time 😁 

cc @segmentio/gateway

### CODE REVIEW

File: cache.go [hunk 1]
```diff
@@ -193,6 +193,23 @@ func (cache *Cache) Remove(key interface{}) bool {
...
+// EvictOldest removes the oldest item from the cache, while also invoking any
+// eviction callback. A bool is returned indicating whether or not an item was
+// removed
+func (cache *Cache) EvictOldest() bool {
+	element := cache.evictionList.Back()
+	if element == nil {
+		return false
+	}
+
+	cache.evictions++
```
Comment (nettofarah) [inline cache.go]: would this operation need to take place within a mutex?

**Review by nettofarah — APPROVED** [review]

File: cache.go (hunk 1 above)
Comment (danielstjules) [reply cache.go]: Good catch, it was previously wrapped with one cause of set

**Review by nettofarah — APPROVED** [review]

File: cache.go [hunk 2]
```diff
@@ -259,6 +269,16 @@ func (cache *Cache) SetMaxAge(maxAge time.Duration) error {
 	return nil
 }
 
+// OnEviction sets the eviction callback
+func (cache *Cache) OnEviction(callback func(key, value interface{})) {
+	cache.onEviction = callback
```
Comment (f2prateek) [inline cache.go]: I don't think this is thread safe.

File: cache.go [hunk 3]
```diff
@@ -259,6 +269,16 @@ func (cache *Cache) SetMaxAge(maxAge time.Duration) error {
...
+}
+
+// OnExpiration sets the expiration callback
+func (cache *Cache) OnExpiration(callback func(key, value interface{})) {
+	cache.onExpiration = callback
```
Comment (f2prateek) [inline cache.go]: I don't think this is thread safe.

**Review by f2prateek — COMMENTED** [review]
It should be ok based if you set it once and forget,  but otherwise I think changing the callbacks isn't thread safe (since they might be set while another goroutine is invoking the old callbacks, and potentially could cause the old callback to be invoked even after OnEviction). Though maybe  it's not a big deal that the behaviour isn't 100% correct here though.

**Comment by danielstjules** [comment]
@f2prateek I can highlight in comments that it's not threadsafe. I'd imagine it's only to be used during cache setup.

**Comment by danielstjules** [comment]
Actually, the only operations that use those callbacks are already guarded by the mutex so this is a quick change :)

**Review by f2prateek — APPROVED** [review]

**Pull request merged**